docs/proposals/decoupling-triggers-and-runtimes.md

```diff
@@ -0,0 +1,117 @@
+# Decoupling triggers and runtimes
+
+## Definition of the problem
+Currently for each new runtime we need to add a container image per trigger. We should design a runtime abstraction. So that:
+ - Triggers can be added (http or event or something else).
+ - These can be in a single language i.e golang
+ - Runtimes can be added more easily
+ - One function can be triggered by more than one trigger source
+
+We need to define interface between trigger container and runtime. What type of protocol to use to pass the request and response.
+
+### **Warning**
+
+Changing the interface between triggers (currently embed in the runtime container) and functions will cause a breaking change. Functions working with previous versions of Kubeless may not work depending on the format of the interface chosen.
+
+## User POV
+From the users point of view we would support:
+
+```bash
+# As today, deploy runtime + trigger
+kubeless function deploy func --trigger-http [...]
+
+# Deploy the runtime without a trigger
+kubeless function deploy func [...]
+
+# Add a trigger linking it to a function
+kubeless trigger add http --path /func func
+kubeless trigger add kafka_topic --topic s3 func
+```
+
+Note that splitting the trigger type in different "verbs" allow us to easily have flags per trigger type.
+Disclaimer: we would need to define possible flags for each trigger
+Open question: Do we want to implement a relation 1:N between triggers and functions? (one trigger could call several functions)
+
+## Suggested architecture approach
+
+In order to split runtimes and triggers we need a standalone entity that handles each trigger type. For the moment, we will assume that the interface protocol between the trigger and the runtime will be HTTP (discussed later). 
+
+So far we can identify two types of trigger, we will call them _trigger controllers_:
+ - HTTP Trigger: This trigger should redirect HTTP(s) requests from/to the _runtime_. For doing so we can use an Ingress Controller. That way, we can translate each _HTTP Trigger_ for an Ingress Rule.
+ - Kafka Topic Trigger: This trigger should translate topic messages to HTTP request. This way runtimes can have an unique interface, regardless of its trigger. Each _kafka trigger_ can be translated to a Kafka consumer.
+
```

Comment:
I would think of trigger controllers as actual k8s controllers, as triggers should be represented by custom resources on their own. The trigger controllers could then listen for them and create an ingress or a kafka message consumer or whatwever else is needed.

**Open questions to me would be:**
- Should each trigger type have its own CRD, or should there be a common CRD? The different trigger types could still get their own annotations for the controller to look for.
- Should there be a custom resource for each event (e.g. a topic, an HTTP endpoint etc.) and for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.

Reply:
>I would think of trigger controllers as actual k8s controllers, as triggers should be represented by custom resources on their own.

+1. IMO, Function, Trigger and Runtime should be the first class citizen/entity of Kubeless. Each of them should have a CRD. Invoication type, topic name etc can be attribute of the Trigger.

>for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.

+1 for rule type abstraction to associate Function and Trigger. This would also enable n:m association between function and triggers.

Some thing to note though is these changes will be breaking changes.

Reply:
> for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.
One Comment here, as we might have to treat things different. On OpenWhisk they have an API Gateway to expose functions via http. The API Gateway creates an 1:1 realtion between the function and the endpoint (no rules are used here). Triggers are used for more high level async integrations like messaging or chat applications. 

From my point of view, it makes sense to have a trigger + rules concept for async processing. For sync processing (using http) I would like to keep it as simple as possible and reduce the complexity to reduce the response time.

Reply:
For the moment let's take the simple approach: A single CRD per trigger. Each instance of this CRD will contain the references for the functions it should trigger. The trigger controller would create then Ingress rules, Kubernetes services, Kafka consumers or whatever is needed in order to call the function with the less latency possible.